pkg/agent/multicast/mcast_controller.go

```diff
@@ -465,6 +486,107 @@ func (c *Controller) addOrUpdateGroupEvent(e *mcastGroupEvent) {
 	}
 }
 
+func (c *Controller) memberChanged(e interface{}) {
+	podEvent := e.(types.PodUpdate)
+	namespace, name := podEvent.PodNamespace, podEvent.PodName
+	containerID := podEvent.ContainerID
+	interfaceName := util.GenerateContainerInterfaceName(name, namespace, containerID)
+	iface, ok := c.ifaceStore.GetInterfaceByName(interfaceName)
+	klog.Infof("memberChanged: %+v", podEvent.IsAdd)
+	if podEvent.IsAdd {
+		if ok {
+			g := &mcastGroupEvent{
+				eType: groupJoin,
+				iface: iface,
+			}
+			c.queryGroupCh <- g
+		}
+	} else {
+		if ok {
+			g := &mcastGroupEvent{
+				eType: groupLeave,
+				iface: iface,
+			}
+			c.queryGroupCh <- g
+		}
+	}
+}
+
+func (c *Controller) syncQueryGroup(stopCh <-chan struct{}) {
+	for {
+		select {
+		case event := <-c.queryGroupCh:
+			klog.Info("resync query group")
+			if c.queryGroupId == 0 {
+				klog.Infof("c.queryGroupId == 0 %v", c.queryGroupId == 0)
+				return
+			}
+			if event.eType == groupJoin {
+				if event.iface != nil {
+					c.queryGroupMemberLock.Lock()
```

Comment:
If it is a single goroutine, we need not lock?

Reply:
removed